kernel/src/mem.rs

```diff
@@ -31,34 +35,43 @@ use crate::process::ProcessId;
 /// within a single process' address space for the entire lifetime
 /// `'a`. No mutable slice pointing to an overlapping memory region
 /// may exist over the entire lifetime `'a`.
-unsafe fn raw_appslice_to_slice<'a>(ptr: *const u8, len: usize) -> &'a [u8] {
-    use core::ptr::NonNull;
-    use core::slice::from_raw_parts;
-
-    // Rust has very strict requirements on pointer validity[1] which
-    // also in part apply to accesses of length 0. We allow an
-    // application to supply arbitrary pointers if the buffer length is
-    // 0, but this is not allowed for Rust slices. For instance, a null
-    // pointer is _never_ valid, not even for accesses of size zero.
-    //
-    // To get a pointer which does not point to valid (allocated) memory, but
-    // is safe to construct for accesses of size zero, we must call
-    // NonNull::dangling(). The resulting pointer is guaranteed to be well-aligned
-    // and uphold the guarantees required for accesses of size zero.
-    //
-    // [1]: https://doc.rust-lang.org/core/ptr/index.html#safety
-    match len {
-        0 => from_raw_parts(NonNull::<u8>::dangling().as_ptr(), 0),
-        _ => from_raw_parts(ptr, len),
-    }
+unsafe fn raw_processbuf_to_roprocessslice<'a>(
+    ptr: *const u8,
+    len: usize,
+) -> &'a ReadableProcessSlice {
+    // Transmute a reference to a slice of Cell<u8>s into a reference
+    // to a ReadableProcessSlice. This is possible as
+    // ReadableProcessSlice is a #[repr(transparent)] wrapper around a
+    // [ReadableProcessByte], which is a #[repr(transparent)] wrapper
+    // around a [Cell<u8>], which is a #[repr(transparent)] wrapper
+    // around an [UnsafeCell<u8>], which finally #[repr(transparent)]
+    // wraps a [u8]
+    core::mem::transmute::<&[u8], &ReadableProcessSlice>(
+        // Rust has very strict requirements on pointer validity[1] which
+        // also in part apply to accesses of length 0. We allow an
+        // application to supply arbitrary pointers if the buffer length is
+        // 0, but this is not allowed for Rust slices. For instance, a null
+        // pointer is _never_ valid, not even for accesses of size zero.
+        //
+        // To get a pointer which does not point to valid (allocated) memory, but
+        // is safe to construct for accesses of size zero, we must call
+        // NonNull::dangling(). The resulting pointer is guaranteed to be well-aligned
+        // and uphold the guarantees required for accesses of size zero.
+        //
+        // [1]: https://doc.rust-lang.org/core/ptr/index.html#safety
+        match len {
+            0 => core::slice::from_raw_parts(core::ptr::NonNull::<u8>::dangling().as_ptr(), 0),
+            _ => core::slice::from_raw_parts(ptr, len),
+        },
+    )
```

Comment:
I don't think `transmute` is needed here. I think something like the following will work:

```rust
// addr should be inferred as a *const ReadableProcessSlice
let addr = match len {
    0 => core::ptr::NonNull::dangling().as_ptr(),
    _ => ptr as *const _,
};
core::slice::from_raw_parts(addr, len)
```

The same applies to `raw_processbuf_to_rwprocessslice`.

Reply:
I'm afraid I don't quite understand. If we were to pass in a `*const ReadableProcessSlice` in `core::slice::from_raw_parts`, wouldn't we get a slice of `ReadableProcessSlice`s? This is also what the compiler appears to do:

```
error[E0308]: mismatched types
  --> kernel/src/mem.rs:73:2
   |
47 | ) -> &'a ReadableProcessSlice {
   |      ------------------------ expected `&'a ReadableProcessSlice` because of return type
...
73 |     core::slice::from_raw_parts(addr, len)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ReadableProcessSlice`, found slice
   |
   = note: expected reference `&'a ReadableProcessSlice`
              found reference `&[_]`
```

I think we need to use `transmute` here, as we have a proper Rust slice and want to convert it to a type which -- under the hood -- is still a proper Rust slice, but is a newtype and therefore can't be treated as one.

Reply:
Oh, I see what's happening. Yes, you're right, I think you need `transmute` here.